include/Spectra/JDSymEigsBase.h

```diff
@@ -0,0 +1,185 @@
+#include "Util/RitzPairs.h"
+
+namespace Spectra {
+
+template <typename Scalar,
+          typename OpType>
+class JDSymEigsBase
+{
+private:
+    using Index = Eigen::Index;
+    using Matrix = Eigen::Matrix<Scalar, Eigen::Dynamic, Eigen::Dynamic>;
+    using Vector = Eigen::Matrix<Scalar, Eigen::Dynamic, 1>;
+    using Array = Eigen::Array<Scalar, Eigen::Dynamic, 1>;
+    using BoolArray = Eigen::Array<bool, Eigen::Dynamic, 1>;
+    using MapMat = Eigen::Map<Matrix>;
+    using MapVec = Eigen::Map<Vector>;
+    using MapConstVec = Eigen::Map<const Vector>;
+
+public:
+    JDSymEigsBase(OpType& op, Index nev) :
+        matrix_operator_(op),
+        operator_dimension_(op.rows()),
+        number_eigenvalues_(nev),
+        max_search_space_size_(10 * number_eigenvalues_),
+        initial_search_space_size_(2 * number_eigenvalues_)
+    {
+        check_argument();
+    }
+
+    ///
+    /// Sets the Maxmium SearchspaceSize after which is deflated
+    ///
+    void setMaxSearchspaceSize(Index max_search_space_size)
+    {
+        max_search_space_size_ = max_search_space_size;
+    }
+
+    ///
+    /// Sets the Initial SearchspaceSize for Ritz values
+    ///
+    void setInitialSearchspaceSize(Index initial_search_space_size)
+    {
+        initial_search_space_size_ = initial_search_space_size;
+    }
+
+    ///
+    /// Virtual destructor
+    ///
+    virtual ~JDSymEigsBase() {}
+
+    ///
+    /// Returns the status of the computation.
+    /// The full list of enumeration values can be found in \ref Enumerations.
+    ///
+    CompInfo info() const { return info_; }
+
+    ///
+    /// Returns the number of iterations used in the computation.
+    ///
+    Index num_iterations() const { return niter_; }
+
+    Vector eigenvalues() const { return ritz_pairs_.RitzValues().head(number_eigenvalues_); }
+
+    Matrix eigenvectors() const { return ritz_pairs_.RitzVectors().leftCols(number_eigenvalues_); }
+
+    ///
+    /// Initializes the solver by providing an initial search space.
+    ///
+    /// \param init_resid Matrix containing initial search space.
+    ///
+    ///
+    void init(const Eigen::Ref<const Matrix>& init_space)
+    {
+        niter_ = 0;
+        search_space_.BasisVectors() = init_space;
+    }
+
+    ///
+    /// Initializes the solver by providing  initial search sapce.
+    ///
+    /// Depending n the method this search space is initialized in a different way
+    ///
+    void init()
+    {
+        Matrix intial_space = SetupInitialSearchSpace();
+        //TODO orthogonalize
+        init(intial_space);
+    }
+
+protected:
+    virtual Matrix SetupInitialSearchSpace(SortRule selection) = 0;
+
+    virtual Matrix CalculateCorrectionVector() const = 0;
+    const OpType& matrix_operator_;  // object to conduct matrix operation,
+                                     // e.g. matrix-vector product
+
+    Index niter_ = 0;
+    const Index operator_dimension_;  // dimension of matrix A
```

Comment:
hmm this should be part of the matvec classes.

Reply:
indeed, now I use the `rows` (or `cols`) methods